File: frameworks/pytorch/skip_elman.py
# Copyright 2020 LMNT, Inc. All Rights Reserved.
# Modified 2024 for SkipElman
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

"""SkipElman: Elman RNN with skip connection and separate input/hidden biases.

Formula:
  z = sigmoid(Wz @ x + Rz @ h + bx_z + bh_z)   # update gate
  a = tanh(Wa @ x + Ra @ h + bx_a + bh_a)      # candidate
  h_new = z * h + (1-z) * a                     # skip connection!

Simpler than GRU (2 components, no reset gate) but retains gradient highway.
"""


import haste_pytorch_lib as LIB
import torch
import torch.nn as nn
import torch.nn.functional as F

from .base_rnn import BaseRNN


__all__ = [
    'SkipElman'
]


#@torch.jit.script
def SkipElmanScript(
    training: bool,
    zoneout_prob: float,
    input,
    h0,
    kernel,
    recurrent_kernel,
    input_bias,
    hidden_bias,
    zoneout_mask):
  """Pure PyTorch fallback for CPU."""
  time_steps = input.shape[0]
  batch_size = input.shape[1]
  hidden_size = recurrent_kernel.shape[0]

  h = [h0]
  Wx = input @ kernel  # [T, B, H*2]
  for t in range(time_steps):
    Rh = h[t] @ recurrent_kernel  # [B, H*2]
    # Split into z (update gate) and a (candidate)
    vx = torch.chunk(Wx[t], 2, 1)
    vh = torch.chunk(Rh, 2, 1)
    vbx = torch.chunk(input_bias, 2, 0)
    vbh = torch.chunk(hidden_bias, 2, 0)

    z = torch.sigmoid(vx[0] + vh[0] + vbx[0] + vbh[0])
    a = torch.tanh(vx[1] + vh[1] + vbx[1] + vbh[1])

    # Skip connection: h_new = z * h + (1-z) * a
    h.append(z * h[t] + (1 - z) * a)
    if zoneout_prob:
      if training:
        h[-1] = (h[-1] - h[-2]) * zoneout_mask[t] + h[-2]
      else:
        h[-1] = zoneout_prob * h[-2] + (1 - zoneout_prob) * h[-1]

  h = torch.stack(h)
  return h


class SkipElmanFunction(torch.autograd.Function):
  @staticmethod
  def forward(ctx, training, zoneout_prob, *inputs):
    h, cache = LIB.skip_elman_forward(training, zoneout_prob, *inputs)
    ctx.save_for_backward(inputs[0], *inputs[2:], h, cache)
    ctx.mark_non_differentiable(inputs[-1])
    ctx.training = training
    return h

  @staticmethod
  def backward(ctx, grad_h):
    if not ctx.training:
      raise RuntimeError('SkipElman backward can only be called in training mode')

    saved = [*ctx.saved_tensors]
    saved[0] = saved[0].permute(2, 0, 1).contiguous()
    saved[1] = saved[1].permute(1, 0).contiguous()
    saved[2] = saved[2].permute(1, 0).contiguous()
    grads = LIB.skip_elman_backward(*saved, grad_h.contiguous())
    return (None, None, *grads, None)


class SkipElman(BaseRNN):
  """
  SkipElman layer: Elman RNN with skip connection.

  This layer implements a simplified GRU without the reset gate:
    z = sigmoid(Wz @ x + Rz @ h + bx_z + bh_z)   # update gate
    a = tanh(Wa @ x + Ra @ h + bx_a + bh_a)      # candidate
    h_new = z * h + (1-z) * a                     # skip connection

  The skip connection (z * h term) provides a gradient highway similar to GRU,
  enabling better long-range learning compared to standard Elman.

  Key differences from GRU:
    - No reset gate (simpler, fewer parameters)
    - Separate input and hidden biases (like cuDNN)

  This layer has built-in support for DropConnect and Zoneout regularization.
  """

  def __init__(self,
      input_size,
      hidden_size,
      batch_first=False,
      dropout=0.0,
      zoneout=0.0,
      return_state_sequence=False):
    """
    Initialize the parameters of the SkipElman layer.

    Arguments:
      input_size: int, the feature dimension of the input.
      hidden_size: int, the feature dimension of the output.
      batch_first: (optional) bool, if `True`, then the input and output
        tensors are provided as `(batch, seq, feature)`.
      dropout: (optional) float, sets the dropout rate for DropConnect
        regularization on the recurrent matrix.
      zoneout: (optional) float, sets the zoneout rate for Zoneout
        regularization.
      return_state_sequence: (optional) bool, if `True`, the forward pass will
        return the entire state sequence instead of just the final state.

    Variables:
      kernel: input projection weight matrix. Dimensions (input_size, hidden_size * 2)
        with `z,a` gate layout. Initialized with Xavier uniform.
      recurrent_kernel: recurrent projection weight matrix. Dimensions
        (hidden_size, hidden_size * 2) with `z,a` gate layout. Initialized orthogonal.
      input_bias: input projection bias vector. Dimensions (hidden_size * 2).
      hidden_bias: recurrent projection bias vector. Dimensions (hidden_size * 2).
    """
    super().__init__(input_size, hidden_size, batch_first, zoneout, return_state_sequence)

    if dropout < 0 or dropout > 1:
      raise ValueError('SkipElman: dropout must be in [0.0, 1.0]')
    if zoneout < 0 or zoneout > 1:
      raise ValueError('SkipElman: zoneout must be in [0.0, 1.0]')

    self.dropout = dropout

    # SkipElman has 2 components (z, a) instead of GRU's 3 (z, r, h)
    self.kernel = nn.Parameter(torch.empty(input_size, hidden_size * 2))
    self.recurrent_kernel = nn.Parameter(torch.empty(hidden_size, hidden_size * 2))
    self.input_bias = nn.Parameter(torch.empty(hidden_size * 2))
    self.hidden_bias = nn.Parameter(torch.empty(hidden_size * 2))
    self.reset_parameters()

  def reset_parameters(self):
    """Resets this layer's parameters to their initial values."""
    hidden_size = self.hidden_size
    # Initialize each gate separately
    for i in range(2):
      nn.init.xavier_uniform_(self.kernel[:, i*hidden_size:(i+1)*hidden_size])
      nn.init.orthogonal_(self.recurrent_kernel[:, i*hidden_size:(i+1)*hidden_size])
    nn.init.zeros_(self.input_bias)
    nn.init.zeros_(self.hidden_bias)

  def forward(self, input, state=None, lengths=None):
    """
    Runs a forward pass of the SkipElman layer.

    Arguments:
      input: Tensor, a batch of input sequences to pass through the SkipElman.
        Dimensions (seq_len, batch_size, input_size) if `batch_first` is
        `False`, otherwise (batch_size, seq_len, input_size).
      state: (optional) Tensor, the initial hidden state. Dimensions
        (1, batch_size, hidden_size).
      lengths: (optional) Tensor, list of sequence lengths for each batch
        element. Dimension (batch_size).

    Returns:
      output: Tensor, the output of the SkipElman layer. Dimensions
        (seq_len, batch_size, hidden_size) if `batch_first` is `False` (default)
        or (batch_size, seq_len, hidden_size) if `batch_first` is `True`.
      h_n: the hidden state for the last sequence item. Dimensions
        (1, batch_size, hidden_size).
    """
    input = self._permute(input)
    state_shape = [1, input.shape[1], self.hidden_size]
    h0 = self._get_state(input, state, state_shape)
    h = self._impl(input, h0[0], self._get_zoneout_mask(input))
    state = self._get_final_state(h, lengths)
    output = self._permute(h[1:])
    return output, state

  def _impl(self, input, state, zoneout_mask):
    if self._is_cuda():
      return SkipElmanFunction.apply(
          self.training,
          self.zoneout,
          input.contiguous(),
          state.contiguous(),
          self.kernel.contiguous(),
          F.dropout(self.recurrent_kernel, self.dropout, self.training).contiguous(),
          self.input_bias.contiguous(),
          self.hidden_bias.contiguous(),
          zoneout_mask.contiguous())
    else:
      return SkipElmanScript(
          self.training,
          self.zoneout,
          input.contiguous(),
          state.contiguous(),
          self.kernel.contiguous(),
          F.dropout(self.recurrent_kernel, self.dropout, self.training).contiguous(),
          self.input_bias.contiguous(),
          self.hidden_bias.contiguous(),
          zoneout_mask.contiguous())

File: frameworks/pytorch/skip_elman.cc
// Copyright 2020 LMNT, Inc. All Rights Reserved.
// Modified 2024 for SkipElman
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

#include <ATen/cuda/CUDAContext.h>
#include <c10/cuda/CUDAGuard.h>
#include <torch/extension.h>
#include <vector>

#include "haste.h"
#include "support.h"

namespace {

using haste::v0::skip_elman::ForwardPass;
using haste::v0::skip_elman::BackwardPass;

using torch::Tensor;

std::vector<Tensor> skip_elman_forward(
    bool training,
    float zoneout_prob,
    Tensor x,
    Tensor h0,
    Tensor kernel,
    Tensor recurrent_kernel,
    Tensor input_bias,
    Tensor hidden_bias,
    Tensor zoneout_mask) {
  const auto time_steps = x.size(0);
  const auto batch_size = x.size(1);
  const auto input_size = x.size(2);
  const auto hidden_size = recurrent_kernel.size(0);
  const bool has_zoneout = zoneout_prob && zoneout_mask.size(0);

  CHECK_INPUT(x);
  CHECK_INPUT(h0);
  CHECK_INPUT(kernel);
  CHECK_INPUT(recurrent_kernel);
  CHECK_INPUT(input_bias);
  CHECK_INPUT(hidden_bias);
  CHECK_INPUT(zoneout_mask);

  const auto options = x.options();
  const at::cuda::CUDAGuard guard(options.device_index());
  // SkipElman has 2 components (z, a) vs GRU's 3 (z, r, g)
  Tensor output = torch::empty({ time_steps + 1, batch_size, hidden_size }, options);
  Tensor cache = torch::empty({ time_steps, batch_size, hidden_size * 2 }, options);
  Tensor tmp_Wx = torch::empty({ time_steps, batch_size, hidden_size * 2 }, options);
  Tensor tmp_Rh = torch::empty({ batch_size, hidden_size * 2 }, options);

  output[0] = h0;

  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, x.scalar_type(), "skip_elman_forward", ([&] {
    ForwardPass<typename native_type<scalar_t>::T> forward(
        training,
        batch_size,
        input_size,
        hidden_size,
        at::cuda::getCurrentCUDABlasHandle(),
        at::cuda::getCurrentCUDAStream());

    forward.Run(
        time_steps,
        ptr<scalar_t>(kernel),
        ptr<scalar_t>(recurrent_kernel),
        ptr<scalar_t>(input_bias),
        ptr<scalar_t>(hidden_bias),
        ptr<scalar_t>(x),
        ptr<scalar_t>(output),
        ptr<scalar_t>(cache),
        ptr<scalar_t>(tmp_Wx),
        ptr<scalar_t>(tmp_Rh),
        has_zoneout ? zoneout_prob : 0.0f,
        has_zoneout ? ptr<scalar_t>(zoneout_mask) : nullptr);
  }));

  return { output, cache };
}

std::vector<Tensor> skip_elman_backward(
    Tensor x_t,
    Tensor kernel_t,
    Tensor recurrent_kernel_t,
    Tensor input_bias,
    Tensor hidden_bias,
    Tensor zoneout_mask,
    Tensor h,
    Tensor cache,
    Tensor dh_new) {
  const auto input_size = x_t.size(0);
  const auto time_steps = x_t.size(1);
  const auto batch_size = x_t.size(2);
  const auto hidden_size = recurrent_kernel_t.size(1);
  const bool has_zoneout = !!zoneout_mask.size(0);

  CHECK_INPUT(x_t);
  CHECK_INPUT(kernel_t);
  CHECK_INPUT(recurrent_kernel_t);
  CHECK_INPUT(input_bias);
  CHECK_INPUT(hidden_bias);
  CHECK_INPUT(h);
  CHECK_INPUT(cache);
  CHECK_INPUT(dh_new);
  CHECK_INPUT(zoneout_mask);

  const auto options = x_t.options();
  const at::cuda::CUDAGuard guard(options.device_index());
  // SkipElman has 2 components
  Tensor dx = torch::empty({ time_steps, batch_size, input_size }, options);
  Tensor dW = torch::zeros({ input_size, hidden_size * 2 }, options);
  Tensor dR = torch::zeros({ hidden_size, hidden_size * 2 }, options);
  Tensor dbx = torch::zeros({ hidden_size * 2 }, options);
  Tensor dbh = torch::zeros({ hidden_size * 2 }, options);
  Tensor dh = torch::zeros({ batch_size, hidden_size }, options);
  Tensor dp = torch::empty({ time_steps, batch_size, hidden_size * 2 }, options);
  Tensor dq = torch::empty({ time_steps, batch_size, hidden_size * 2 }, options);

  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, x_t.scalar_type(), "skip_elman_backward", ([&] {
    BackwardPass<typename native_type<scalar_t>::T> backward(
        batch_size,
        input_size,
        hidden_size,
        at::cuda::getCurrentCUDABlasHandle(),
        at::cuda::getCurrentCUDAStream());

    backward.Run(
        time_steps,
        ptr<scalar_t>(kernel_t),
        ptr<scalar_t>(recurrent_kernel_t),
        ptr<scalar_t>(input_bias),
        ptr<scalar_t>(hidden_bias),
        ptr<scalar_t>(x_t),
        ptr<scalar_t>(h),
        ptr<scalar_t>(cache),
        ptr<scalar_t>(dh_new),
        ptr<scalar_t>(dx),
        ptr<scalar_t>(dW),
        ptr<scalar_t>(dR),
        ptr<scalar_t>(dbx),
        ptr<scalar_t>(dbh),
        ptr<scalar_t>(dh),
        ptr<scalar_t>(dp),
        ptr<scalar_t>(dq),
        has_zoneout ? ptr<scalar_t>(zoneout_mask) : nullptr);
  }));

  return { dx, dh, dW, dR, dbx, dbh };
}

}  // anonymous namespace

void skip_elman_init(py::module& m) {
  m.def("skip_elman_forward", &skip_elman_forward, "SkipElman forward", py::call_guard<py::gil_scoped_release>());
  m.def("skip_elman_backward", &skip_elman_backward, "SkipElman backward", py::call_guard<py::gil_scoped_release>());
}
